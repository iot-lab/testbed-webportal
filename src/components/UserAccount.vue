<template>
<div class="container">
    <h2><i class="fa fa-fw fa-user" aria-hidden="true"></i> My Account</h2>
    <div class="row">
      <div class="col-md-3">
        <div class="list-group" id="list-tab" role="tablist">
          <a class="list-group-item list-group-item-action active" id="list-profile-list" data-toggle="list" href="#list-profile" role="tab" aria-controls="profile"><i class="fa fa-fw fa-address-card" aria-hidden="true"></i> Profile</a>
          <a class="list-group-item list-group-item-action" id="list-password-list" data-toggle="list" href="#list-password" role="tab" aria-controls="password"><i class="fa fa-fw fa-unlock-alt" aria-hidden="true"></i> Password</a>
          <a class="list-group-item list-group-item-action" id="list-sshkeys-list" data-toggle="list" href="#list-sshkeys" role="tab" aria-controls="sshkeys"><i class="fa fa-fw fa-key" aria-hidden="true"></i> SSH Keys</a>
          <a class="list-group-item list-group-item-action" id="list-delete-list" data-toggle="list" href="#list-delete" role="tab" aria-controls="delete"><i class="fa fa-fw fa-trash-o" aria-hidden="true"></i> Delete account</a>
        </div>
      </div>
      <div class="col-9">
        <div class="tab-content" id="nav-tabContent">
          <div class="tab-pane fade show active" id="list-profile" role="tabpanel" aria-labelledby="list-profile-list">
            
            <h5><i class="fa fa-fw fa-pencil" aria-hidden="true"></i> Edit your profile</h5>
            <form @submit.prevent="saveDetails">
                <div class="row">
                  <div class="col-md">
                    <div class="form-group">
                      <label class="form-control-label">First name</label>
                      <input placeholder="First name" v-model="user.firstName" class="form-control" type="text" required>
                    </div>
                  </div>
                  <div class="col-md">                    
                    <div class="form-group">
                      <label class="form-control-label">Last name</label>
                      <input placeholder="Last name" v-model="user.lastName" class="form-control" type="text" required>
                    </div>
                  </div>
                </div>
                <div class="row">
                  <div class="col-md">
                    <div class="form-group">
                      <label class="form-control-label">Email</label>
                      <input v-model="user.email" class="form-control" type="email" required placeholder="Academic or professional email">
                    </div>
                  </div>
                  <div class="col-md"></div>
                </div>
                <div class="row">
                  <div class="col-md">
                    <div class="form-group">
                      <label class="form-control-label">User category</label>
                      <v-select v-model="user.category" :options="categories" placeholder="Category" :searchable="false" required></v-select>
                    </div>                    
                  </div>
                  <div class="col-md">
                    <div class="form-group">
                      <label class="form-control-label">Organization</label>
                      <input placeholder="Organization" v-model="user.organization" class="form-control" type="text" required>
                    </div>
                  </div>
                </div>
                <div class="row">
                  <div class="col-md">
                    <div class="form-group">
                      <label class="form-control-label">City</label>
                      <input placeholder="City" v-model="user.city" class="form-control" type="text" required>
                    </div>
                  </div>
                  <div class="col-md">
                    <div class="form-group">
                      <label class="form-control-label">Country</label>
                      <v-select :options="countries" v-model="user.country" required placeholder="Country" @keydown.enter.prevent=""></v-select>
                    </div>
                  </div>
                </div>
                <div class="form-group">
                  <label class="form-control-label">Motivations</label>
                  <textarea v-model="user.motivations" class="form-control" rows="5" placeholder="Tell us about you" required></textarea>
                </div>
                <div class="form-group">
                  <div class="alert alert-info form-control">
                    <i class="fa fa-pencil"></i> <b>Tell us about your motivations</b>
                    <ul style="margin: 0">
                      <li>Research domain (Radio communication, networking protocol, distributed applications, &#x2026;)</li>
                      <li>What kind of experiments do you want to run with IoT-LAB?</li>
                      <li>Goal (Verify something existing in large scale, new development, &#x2026;)</li>
                      <li>Experience with sensor networks (n00b, experiments with X platform, former IoT-LAB user, Guru, God)</li>
                    </ul>
                  </div>
                </div>
                <div class="form-group">
                  <button class="btn btn-success" type="submit">Update profile</button>
                  <button class="btn btn-secondary" type="button" @click="reset">Reset</button>
                </div>
            </form>
            
          </div>
          <div class="tab-pane fade" id="list-password" role="tabpanel" aria-labelledby="list-password-list">
            
            <h5><i class="fa fa-fw fa-user" aria-hidden="true"></i> Username <span class="text-muted">{{auth.username}}</span></h5>
            <p class="mb-4">Your username cannot be modified.</p>
            <h5><i class="fa fa-fw fa-unlock-alt" aria-hidden="true"></i> Change your password</h5>
            <div class="row">
              <form class="" @submit.prevent="changePassword" class="col-md-6"> 
                <input v-model="pwd.old" type="password" placeholder="Current password" class="form-control" style="margin-top: 8px;" required>
                <input v-model="pwd.new" type="password" placeholder="New password" class="form-control" style="margin-top: 8px;" required>
                <input v-model="pwd.confirm" type="password" placeholder="Confirm password" class="form-control" style="margin-top: 8px;" required>
                <button class="btn btn-success" type="submit" style="margin-top: 10px;">Change Password</button>
              </form>            
            </div>
          </div>

          <div class="tab-pane fade" id="list-sshkeys" role="tabpanel" aria-labelledby="list-sshkeys-list">

            <h5><i class="fa fa-fw fa-key" aria-hidden="true"></i> SSH keys</h5>
            <form class="" @submit.prevent="saveKeys">
                <ul class="nav nav-tabs" style="border-bottom: 1px solid transparent; position: relative; top: 1px">
                    <li v-for="(key, i) in keys" class="nav-item"> <a :href="'#tab_SSH'+i" :class="{ 'active': i === activeKey }" data-toggle="tab" @click="activeKey = i"
                        class="nav-link">
                        SSH key {{i+1}}
                        <a @click.stop.prevent="delKey(i)" v-show="i === activeKey && keys.length > 1"><i class="fa fa-times-circle"></i></a> </a>
                    </li>
                    <li class="nav-item"><a @click="addKey" class="nav-link"><i class="fa fa-plus"></i></a>
                    </li>
                </ul>
                <div class="tab-content" id="sshkeysTabContent">
                    <div v-for="(key, i) in keys" class="tab-pane" :class="{ 'active': i === activeKey, 'show': i === 0 }"
                    :id="'tab_SSH'+i">
                        <div class="control-group">
                            <div class="controls">
                                <textarea v-model="keys[i]" class="form-control" rows="5"></textarea>
                            </div>
                        </div>
                    </div>
                </div>
                <button class="btn btn-success" type="submit" style="margin-top: 10px;">Update SSH keys</button>
            </form>
          </div>
          <div class="tab-pane fade" id="list-delete" role="tabpanel" aria-labelledby="list-delete-list">
            <p class="lead">Permanently delete your account?</p>
            <p><strong>All your data will be lost</strong></p>
            <button class="btn btn-danger" type="button"><i class="fa fa-trash-o fa-lg" aria-hidden="true"></i> Delete Account</button>
          </div>
        </div>
      </div>
    </div>
</div> <!-- container -->
</template>

<script>
import vSelect from 'vue-select'
import countries from '@/assets/js/countries'
import categories from '@/assets/js/categories'
import {iotlab} from '@/rest'
import {auth} from '@/auth'

export default {
  name: 'UserAccount',
  components: {vSelect},

  data () {
    return {
      user: {},
      keys: [''],
      pwd: {},
      countries: countries,
      categories: categories,
      auth: auth,
      activeKey: 0,
    }
  },

  created () {
    iotlab.getSSHkeys().then(data => { this.keys = data })
    iotlab.getUserInfo().then(data => { this.user = data })
  },

  methods: {
    async reset () {
      this.user = await iotlab.getUserInfo()
    },
    async saveDetails () {
      if (typeof this.user.category === 'object') {
        this.user.category = this.user.category.value
      }
      await iotlab.setUserInfo(this.user)
      this.$notify({text: 'Profile updated', type: 'success'})
    },
    async changePassword () {
      try {
        await iotlab.changePassword(this.pwd.old, this.pwd.new, this.pwd.confirm)
        this.pwd = {}
        this.$notify({text: 'Password successfully changed', type: 'success'})
      } catch (err) {
        this.pwd = {}
        this.$notify({text: 'Failed to change password', type: 'error'})
      }
    },
    async saveKeys () {
      try {
        await iotlab.modifySSHkeys(this.keys)
        this.$notify({text: 'SSH keys updated', type: 'success'})
      } catch (err) {
        this.$notify({text: 'Failed to update SSH keys', type: 'error'})
      }
    },
    addKey () {
      this.keys.push('')
      this.activeKey = this.keys.length - 1
    },
    delKey (index) {
      this.keys.splice(index, 1)
      this.activeKey = Math.min(this.activeKey, this.keys.length - 1)
    },
  },
}
</script>
