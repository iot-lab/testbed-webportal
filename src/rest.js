import axios from 'axios'

const API_URL = `https://${process.env.VUE_APP_IOTLAB_HOST}/api/`

export const iotlab = {

  api: axios.create({baseURL: API_URL, auth: JSON.parse(localStorage.getItem('apiAuth') || '{}')}),

  create (username, password) {
    this.api = axios.create({
      baseURL: API_URL,
      auth: {
        username: username,
        password: password,
      },
    })
  },

  // USERS API

  async signup (user, nbUsers) {
    if (nbUsers) {
      // create multiple user accounts
      await iotlab.api.post(`/users?nbusers=${nbUsers}`, user)
    } else {
      // create single user account
      await iotlab.api.post('/users', user)
    }
  },

  async activateAccount (hash) {
    await iotlab.api.post('/users/activate/', {'hash': hash})
  },

  async deleteUser (login = undefined) {
    if (login) {
      await iotlab.api.delete(`/users/${login}`)
    } else {
      await iotlab.api.delete(`/user`)
    }
  },

  async resetPassword (email) {
    await iotlab.api.post(`/users/reset_password`, {'email': email})
  },

  async changePassword (oldPwd, newPwd, confirmPwd) {
    await iotlab.api.put('/user/password', {
      'old_password': oldPwd,
      'new_password': newPwd,
      'confirm_new_password': confirmPwd,
    })
  },

  async getSSHkeys () {
    return iotlab.api.get('/user/keys').then(resp => resp.data.sshkeys)
  },

  async modifySSHkeys (keys) {
    return iotlab.api.post('/user/keys', {'sshkeys': keys})
  },

  async deleteSSHkey (id) {
    return iotlab.api.delete(`/user/keys/${id}`)
  },

  async getUsers ({status, group, search} = {}) {
    let params = {}
    if (status) {
      params.status = status
    }
    if (group) {
      params.group = group
    }
    if (search) {
      params.search = search
    }
    return iotlab.api.get('/users', {params: params}).then(resp => resp.data)
  },

  async getUserInfo (login = undefined) {
    if (login) {
      return iotlab.api.get(`/users/${login}`).then(resp => resp.data)
    } else {
      return iotlab.api.get('/user').then(resp => resp.data)
    }
  },

  async setUserInfo (user, login = undefined) {
    if (login) {
      return iotlab.api.put(`/users/${login}`, user)
    } else {
      return iotlab.api.put('/user', user)
    }
  },

  async activateUser (login) {
    return iotlab.api.put(`/users/${login}`, {'status': 'active'})
  },

  async deactivateUser (login) {
    return iotlab.api.put(`/users/${login}`, {'status': 'inactive'})
  },

  async setUserGroups (login, groups) {
    return iotlab.api.put(`/users/${login}`, {'groups': groups})
  },

  async sendMail (recipients, subject, message) {
    await iotlab.api.post('/users/email', {
      recipients: recipients,
      subject: subject,
      body: message,
    })
  },

  // EXPERIMENT API

  async getUserExperimentsCount () {
    return iotlab.api.get('/experiments/total').then(resp => resp.data)
  },

  async getAllExperimentsCount (user) {
    if (user) {
      return iotlab.api.get(`/experiments/total/users/${user}`).then(resp => resp.data)
    } else {
      return iotlab.api.get(`/experiments/total/all`).then(resp => resp.data)
    }
  },

  async getAllExperiments ({limit, offset, user, state = 'Terminated,Error,Running,Finishing,Resuming,toError,Waiting,Launching,Hold,toLaunch,toAckReservation,Suspended'} = {}) {
    let params = {}
    if (state) {
      params.state = state
    }
    if (limit) {
      params.limit = Math.min(limit, 500) // IoT-LAB api restrict queries to 500 entries
    }
    if (offset) {
      params.offset = offset
    }
    if (user === '@self') {
      return iotlab.api.get('/experiments', {params: params}).then(resp => resp.data.items)
    } else if (user) {
      return iotlab.api.get(`/experiments/users/${user}`, {params: params}).then(resp => resp.data.items)
    } else {
      return iotlab.api.get('/experiments/all', {params: params}).then(resp => resp.data.items)
    }
  },

  async getRunningExperiments () {
    return iotlab.api.get('/experiments/running').then(resp => resp.data.items)
  },

  async getExperiment (id) {
    return iotlab.api.get(`/experiments/${id}`).then(resp => resp.data)
  },

  async getExperimentArchive (id) {
    return iotlab.api.get(`/experiments/${id}/data`, {responseType: 'arraybuffer'}).then(resp => resp.data)
  },

  async getExperimentDeployment (id) {
    return iotlab.api.get(`/experiments/${id}/deployment`).then(resp => resp.data)
  },

  async stopExperiment (id) {
    return iotlab.api.delete(`/experiments/${id}`)
  },

  async submitExperiment ({type, name, duration, reservation, nodes, firmwareassociations, firmwares, profileassociations, profiles} = {}) {
    const formData = new FormData()
    formData.append('experiment.json', JSON.stringify({
      type: type,
      duration: duration,
      reservation: reservation,
      name: name,
      nodes: nodes,
      firmwareassociations: firmwareassociations,
      profileassociations: profileassociations,
    }))
    // attached firmware names should be unique
    let attachedFirmwares = []
    for (let asso of firmwareassociations || []) {
      if (asso.firmwarename in firmwares && !attachedFirmwares.includes(asso.firmwarename)) {
        formData.append(asso.firmwarename, new Blob([firmwares[asso.firmwarename]], {type: 'application/octet-stream'}), asso.firmwarename)
        attachedFirmwares.push(asso.firmwarename)
      }
    }
    return iotlab.api.post('/experiments', formData).then(resp => resp.data)
  },

  async checkFirmware (fileString) {
    const formData = new FormData()
    formData.append('firmware.bin', fileString)
    return iotlab.api.post('/firmwares/checker', formData).then(resp => resp.data)
  },

  async flashFirmware (id, nodes, fileString) {
    const formData = new FormData()
    formData.append(`experiment${id}.json`, JSON.stringify(nodes))
    formData.append('firmware.bin', new Blob([fileString], {type: 'application/octet-stream'}), 'firmware.bin')
    return iotlab.api.post(`/experiments/${id}/nodes/flash`, formData).then(resp => resp.data)
  },

  async flashResourcesFirmware (id, nodes, name) {
    return iotlab.api.post(`/experiments/${id}/nodes/flash/${name}`, nodes).then(resp => resp.data)
  },

  async updateMonitoring (id, nodes, name) {
    return iotlab.api.post(`/experiments/${id}/nodes/monitoring/${name}`, nodes).then(resp => resp.data)
  },

  async updateExperimentMobilityCircuit (id, nodes, name) {
    return iotlab.api.post(`/experiments/${id}/robots/mobility/${name}`, nodes).then(resp => resp.data)
  },

  async reloadExperiment (id) {
    return iotlab.api.post(`/experiments/${id}/reload`, {}).then(resp => resp.data)
  },

  // USER GROUPS

  async getUserGroups () {
    return iotlab.api.get('/groups').then(resp => resp.data)
  },

  async createGroup (name) {
    return iotlab.api.post(`/groups`, {name: name})
  },

  async deleteGroup (group) {
    return iotlab.api.delete(`/groups/${group}`)
  },

  // OTHER API

  async getSites () {
    return iotlab.api.get('/sites').then(resp => resp.data.items)
  },

  async getSitesDetails () {
    return iotlab.api.get('/sites/details').then(resp => resp.data.items)
  },

  async getNodes () {
    return iotlab.api.get('/nodes').then(resp => resp.data.items)
  },

  async updateNodesProperties () {
    return iotlab.api.put('/nodes')
  },

  async updateNodesPropertiesOldApi () {
    const OLD_API = `https://${process.env.VUE_APP_IOTLAB_HOST}/rest/`
    const oldApi = axios.create({baseURL: OLD_API, auth: JSON.parse(localStorage.getItem('apiAuth') || '{}')})
    return oldApi.post('/admin/nodes')
  },

  async getNodesIds () {
    return iotlab.api.get('/nodes/ids').then(resp => resp.data.items)
  },

  async getExperimentNodes (id) {
    return iotlab.api.get(`/experiments/${id}/nodes`).then(resp => resp.data.items)
  },

  async getExperimentToken (id) {
    return iotlab.api.get(`/experiments/${id}/token`).then(resp => resp.data.token)
  },

  async sendNodesCommand (id, cmd, nodes) {
    // cmd in [start, stop, reset, flash-idle, profile-reset, debug-start, debug-stop]
    return iotlab.api.post(`/experiments/${id}/nodes/${cmd}`, nodes).then(resp => resp.data)
  },

  // STORES (firmwares, profiles, etc)

  async getMobilityCircuits () {
    return iotlab.api.get('/mobilities/circuits').then(resp => resp.data.items)
  },

  async getMobilityCircuit (name) {
    return iotlab.api.get(`/mobilities/circuits/${name}`).then(resp => resp.data)
  },

  async getRobotsSiteMapImage (site) {
    return iotlab.api.get(`/robots/${site}/map/image`, {
      responseType: 'arraybuffer',
      headers: {
        'Access-Control-Allow-Origin': '*',
      },
    }).then(resp => Buffer.from(resp.data, 'binary').toString('base64')
    )
  },

  async getRobotsSiteMapConfig (site) {
    return iotlab.api.get(`/robots/${site}/map/config`).then(resp => resp.data)
  },

  async createMobilityCircuit (circuit) {
    return iotlab.api.post(`/mobilities/circuits`, circuit)
  },

  async updateMobilityCircuit (name, circuit) {
    return iotlab.api.put(`/mobilities/circuits/${name}`, circuit)
  },

  async deleteMobilityCircuit (name) {
    return iotlab.api.delete(`/mobilities/circuits/${name}`)
  },

  async getMonitoringProfiles () {
    return iotlab.api.get('/monitoring').then(resp => resp.data)
  },

  async getMonitoringProfile (name) {
    return iotlab.api.get(`/monitoring/${name}`).then(resp => resp.data)
  },

  async updateMonitoringProfile (name, profile) {
    return iotlab.api.put(`/monitoring/${name}`, profile)
  },

  async createMonitoringProfile (profile) {
    return iotlab.api.post(`/monitoring`, profile)
  },

  async deleteMonitoringProfile (name) {
    return iotlab.api.delete(`/monitoring/${name}`)
  },

  async getFirmwares () {
    return iotlab.api.get('/firmwares').then(resp => resp.data)
  },

  async getFirmware (name) {
    return iotlab.api.get(`/firmwares/${name}`).then(resp => resp.data)
  },

  async getFirmwareFile (name) {
    return iotlab.api.get(`/firmwares/${name}/file`, {responseType: 'arraybuffer'}).then(resp => resp.data)
  },

  async updateFirmware (name, firmware, fileString) {
    const formData = new FormData()
    formData.append(`firmware.json`, JSON.stringify(firmware))
    if (fileString) {
      formData.append('firmware.bin', new Blob([fileString], {type: 'application/octet-stream'}), 'firmware.bin')
    }
    return iotlab.api.put(`/firmwares/${name}`, formData).then(resp => resp.data)
  },

  async createFirmware (firmware, fileString) {
    const formData = new FormData()
    formData.append(`firmware.json`, JSON.stringify(firmware))
    formData.append('firmware.bin', new Blob([fileString], {type: 'application/octet-stream'}), 'firmware.bin')
    return iotlab.api.post(`/firmwares`, formData).then(resp => resp.data)
  },

  async deleteFirmware (name) {
    return iotlab.api.delete(`/firmwares/${name}`)
  },
}
