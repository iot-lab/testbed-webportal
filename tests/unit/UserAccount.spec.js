import { shallowMount, createLocalVue } from '@vue/test-utils'
import VueRouter from 'vue-router'
import UserAccount from '@/views/UserAccount'
import { iotlab } from '@/rest'
jest.mock('@/rest')

const localVue = createLocalVue()
localVue.use(VueRouter)
const router = new VueRouter()

const errors = {
  has: jest.fn(),
  first: jest.fn(),
}

describe('UserAccount.vue', () => {

  it('should match snapshot', () => {
    const wrapper = shallowMount(UserAccount, {localVue, router, mocks: {errors}})
    expect(wrapper.element).toBeDefined()
    expect(wrapper.element).toMatchSnapshot()
  })

})
