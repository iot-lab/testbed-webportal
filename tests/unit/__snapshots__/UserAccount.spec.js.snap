// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserAccount.vue should match snapshot 1`] = `
<div
  class="container mt-3"
>
  <h2>
    <i
      aria-hidden="true"
      class="fa fa-fw fa-user"
    />
     My Account
  </h2>
   
  <div
    class="row"
  >
    <div
      class="col-md-3 mb-4"
    >
      <div
        class="list-group"
        id="list-tab"
        role="tablist"
      >
        <a
          aria-controls="profile"
          class="list-group-item list-group-item-action active"
          data-toggle="list"
          href="#list-profile"
          id="list-profile-list"
          role="tab"
        >
          <i
            aria-hidden="true"
            class="fa fa-fw fa-address-card"
          />
           Profile
        </a>
         
        <a
          aria-controls="password"
          class="list-group-item list-group-item-action"
          data-toggle="list"
          href="#list-password"
          id="list-password-list"
          role="tab"
        >
          <i
            aria-hidden="true"
            class="fa fa-fw fa-unlock-alt"
          />
           Password
        </a>
         
        <a
          aria-controls="sshkeys"
          class="list-group-item list-group-item-action"
          data-toggle="list"
          href="#list-sshkeys"
          id="list-sshkeys-list"
          role="tab"
        >
          <i
            aria-hidden="true"
            class="fa fa-fw fa-key"
          />
           SSH Keys
        </a>
         
        <a
          aria-controls="delete"
          class="list-group-item list-group-item-action"
          data-toggle="list"
          href="#list-delete"
          id="list-delete-list"
          role="tab"
        >
          <i
            aria-hidden="true"
            class="fa fa-fw fa-trash-o"
          />
           Delete account
        </a>
         
        <a
          aria-controls="mailing"
          class="list-group-item list-group-item-action"
          data-toggle="list"
          href="#list-mailing"
          id="list-mailing-list"
          role="tab"
        >
          <i
            aria-hidden="true"
            class="fa fa-fw fa-envelope-o"
          />
           Mailing list
        </a>
      </div>
    </div>
     
    <div
      class="col-md-9"
    >
      <div
        class="tab-content"
        id="nav-tabContent"
      >
        <div
          aria-labelledby="list-profile-list"
          class="tab-pane fade show active"
          id="list-profile"
          role="tabpanel"
        >
          <h5>
            <i
              aria-hidden="true"
              class="fa fa-fw fa-pencil"
            />
             Edit your profile
          </h5>
           
          <form>
            <userform-stub />
             
            <div
              class="form-group"
            >
              <button
                class="btn btn-success"
                type="submit"
              >
                Update profile
              </button>
               
              <button
                class="btn btn-secondary"
                type="button"
              >
                Reset
              </button>
            </div>
          </form>
        </div>
         
        <div
          aria-labelledby="list-password-list"
          class="tab-pane fade"
          id="list-password"
          role="tabpanel"
        >
          <h5>
            <i
              aria-hidden="true"
              class="fa fa-fw fa-user"
            />
             Username 
            <span
              class="text-muted"
            >
              
            </span>
          </h5>
           
          <p
            class="mb-4"
          >
            Your username cannot be modified.
          </p>
           
          <h5>
            <i
              aria-hidden="true"
              class="fa fa-fw fa-unlock-alt"
            />
             Change your password
          </h5>
           
          <div
            class="row"
          >
            <form
              class="col-md-5"
            >
              <input
                class="form-control"
                placeholder="Current password"
                required="required"
                style="margin-top: 8px;"
                type="password"
              />
               
              <input
                class="form-control"
                placeholder="New password"
                required="required"
                style="margin-top: 8px;"
                type="password"
              />
               
              <input
                class="form-control"
                placeholder="Confirm password"
                required="required"
                style="margin-top: 8px;"
                type="password"
              />
               
              <button
                class="btn btn-success"
                style="margin-top: 10px;"
                type="submit"
              >
                Change Password
              </button>
            </form>
             
            <div
              class="col-md-6"
            >
              <div
                class="card mt-2 border-danger text-danger"
                id="policy"
                style="display: none;"
              >
                <div
                  class="card-body"
                >
                  <h6>
                    Password Policy
                  </h6>
                   
                  <ul
                    class="pl-3 mb-0"
                  >
                    <li>
                      one upper case letter [A-Z]
                    </li>
                     
                    <li>
                      three lower case letters [a-z]
                    </li>
                     
                    <li>
                      one digit [0-9]
                    </li>
                     
                    <li>
                      one special character [!@#$%^&*_=+-/]
                    </li>
                     
                    <li>
                      minimum length of eight characters
                    </li>
                  </ul>
                </div>
              </div>
            </div>
          </div>
        </div>
         
        <div
          aria-labelledby="list-sshkeys-list"
          class="tab-pane fade"
          id="list-sshkeys"
          role="tabpanel"
        >
          <h5>
            <i
              aria-hidden="true"
              class="fa fa-fw fa-key"
            />
             SSH keys
          </h5>
           
          <form>
            <sshkeys-stub />
             
            <button
              class="btn btn-success"
              style="margin-top: 10px;"
              type="submit"
            >
              Update SSH keys
            </button>
          </form>
        </div>
         
        <div
          aria-labelledby="list-delete-list"
          class="tab-pane fade"
          id="list-delete"
          role="tabpanel"
        >
          <p
            class="lead"
          >
            Permanently delete your account?
          </p>
           
          <p>
            <strong>
              All your data will be lost
            </strong>
          </p>
           
          <button
            class="btn btn-danger"
            type="button"
          >
            <i
              aria-hidden="true"
              class="fa fa-trash-o fa-lg"
            />
             Delete Account
          </button>
        </div>
         
        <div
          aria-labelledby="list-mailing-list"
          class="tab-pane fade"
          id="list-mailing"
          role="tabpanel"
        >
          <h5
            class="mb-3"
          >
            <i
              aria-hidden="true"
              class="fa fa-envelope-o"
            />
             Mailing list subscription
          </h5>
           
          <p>
            You can manage your subscription to the mailing list for FIT IoT-LAB users from this page:
            <br />
             
            <a
              href="https://lists.gforge.inria.fr/mailman/listinfo/senslab-users"
            >
              https://lists.gforge.inria.fr/mailman/listinfo/senslab-users
            </a>
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;
